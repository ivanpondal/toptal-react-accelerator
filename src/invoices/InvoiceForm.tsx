import {
  Alert,
  Autocomplete,
  Box,
  Button,
  Divider,
  Grid,
  IconButton,
  Stack,
  TextField,
  Typography,
} from "@mui/material";
import { DesktopDatePicker, LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDateFns } from "@mui/x-date-pickers/AdapterDateFns";
import { Controller, useFieldArray, useForm } from "react-hook-form";
import * as yup from "yup";
import FormTextField, {
  ControlledFormTextField,
} from "../components/form-wrappers";
import LoadingButton from "../components/LoadingButton";
import ClearIcon from "@mui/icons-material/Clear";
import { yupResolver } from "@hookform/resolvers/yup";
import { useEffect } from "react";

const schema = yup.object({
  invoiceDate: yup.date().required(),
  invoiceDueDate: yup.date().min(yup.ref("invoiceDate")).required(),
  invoiceNumber: yup.string().min(3).required(),
  invoiceProjectCode: yup
    .string()
    .nullable()
    .notRequired()
    .min(3)
    .transform((value) => (!!value ? value : null)),
  invoiceClientId: yup.string().required(),
  items: yup.array().of(
    yup.object().shape({
      description: yup.string().min(3).required(),
      value: yup.number().moreThan(0).required(),
    })
  ),
  total: yup.number().moreThan(0).required(),
});

export type InvoiceDetailsFormData = {
  invoiceDate: Date;
  invoiceDueDate: Date;
  invoiceNumber: string;
  invoiceProjectCode?: string;
  invoiceClientId: string;
  items: Array<{
    description: string;
    value: number;
  }>;
  total: number;
};

export type InvoiceDetailsFormProps = {
  onSubmitRequest: (values: InvoiceDetailsFormData) => Promise<boolean>;
  clientNames: { [key: string]: { id: string; label: string } };
  isUpdateForm?: boolean;
  loading?: boolean;
  errorMessage?: string;
  successMessage?: string;
  invoice?: InvoiceDetailsFormData;
};

export default function InvoiceForm(props: InvoiceDetailsFormProps) {
  const {
    clientNames,
    loading,
    errorMessage,
    successMessage,
    onSubmitRequest,
    invoice,
    isUpdateForm = false,
  } = props;

  const {
    register,
    handleSubmit,
    formState: { errors },
    reset,
    control,
    watch,
    setValue,
  } = useForm<InvoiceDetailsFormData>({
    resolver: yupResolver(schema),
    defaultValues: {
      invoiceNumber: "",
      invoiceProjectCode: "",
      items: [{ description: "", value: 0 }],
    },
  });

  const { fields, append, remove } = useFieldArray({
    control,
    name: "items",
  });

  const total = watch("items")
    .map((item) => item.value)
    .map((value) => Number(value))
    .filter((value) => !Number.isNaN(value))
    .filter((value) => value > 0)
    .reduce((prev, curr) => prev + curr, 0);

  setValue("total", total);

  useEffect(() => {
    if (invoice) {
      reset(invoice);
    }
  }, [reset, invoice]);

  return (
    <Box
      component="form"
      onSubmit={handleSubmit(async (data) => {
        const success = await onSubmitRequest(data);
        if (success && !isUpdateForm) {
          reset();
        }
      })}
      noValidate
      sx={{ mt: 1 }}
    >
      {successMessage && (
        <Alert data-test="form-success" severity="success">
          {successMessage}
        </Alert>
      )}

      {errorMessage && (
        <Alert data-test="form-error" severity="error">
          {errorMessage}
        </Alert>
      )}

      <LocalizationProvider dateAdapter={AdapterDateFns}>
        <Stack direction="row" spacing={3} sx={{ mt: 2, mb: 1 }}>
          <Controller
            name="invoiceDate"
            control={control}
            defaultValue={new Date()}
            render={({ field }) => (
              <DesktopDatePicker
                label="Date"
                inputFormat="MM/dd/yyyy"
                {...field}
                renderInput={(params) => {
                  return (
                    <TextField
                      margin="normal"
                      fullWidth
                      helperText={
                        errors.invoiceDate && (
                          <span data-test="invoice-date-error">
                            {errors.invoiceDate?.message}
                          </span>
                        )
                      }
                      inputProps={{ "data-test": "invoice-date" }}
                      sx={{ m: 0 }}
                      {...params}
                      error={!!errors.invoiceDate}
                      disabled={loading}
                    />
                  );
                }}
              />
            )}
          />

          <Controller
            name="invoiceDueDate"
            control={control}
            defaultValue={new Date()}
            render={({ field }) => (
              <DesktopDatePicker
                label="Due Date"
                inputFormat="MM/dd/yyyy"
                {...field}
                renderInput={(params) => {
                  return (
                    <TextField
                      margin="normal"
                      fullWidth
                      helperText={
                        errors.invoiceDueDate && (
                          <span data-test="invoice-due-date-error">
                            {errors.invoiceDueDate?.message}
                          </span>
                        )
                      }
                      inputProps={{ "data-test": "invoice-due-date" }}
                      sx={{ m: 0 }}
                      {...params}
                      error={!!errors.invoiceDueDate}
                      disabled={loading}
                    />
                  );
                }}
              />
            )}
          />
        </Stack>
      </LocalizationProvider>

      <ControlledFormTextField
        id="invoiceNumber"
        control={control}
        label="Number"
        errorField={errors.invoiceNumber}
        dataTestId="invoice-number"
        disabled={loading}
        required
      />

      <ControlledFormTextField
        id="invoiceProjectCode"
        control={control}
        label="Project Code"
        errorField={errors.invoiceProjectCode}
        dataTestId="invoice-project-code"
        disabled={loading}
      />

      <Controller
        name="invoiceClientId"
        control={control}
        render={({ field }) => (
          <Autocomplete
            {...field}
            disablePortal
            sx={{ mt: 2, mb: 1 }}
            options={Object.values(clientNames)}
            onChange={(_, data) => field.onChange(data?.id)}
            value={field.value ? clientNames[field.value] : null}
            renderInput={(params) => {
              let { inputProps, ...restParams } = params;
              return (
                <TextField
                  margin="normal"
                  sx={{ m: 0 }}
                  {...restParams}
                  inputProps={{
                    "data-test": "invoice-company-id",
                    ...inputProps,
                  }}
                  helperText={
                    errors.invoiceClientId && (
                      <span data-test="invoice-company-id-error">
                        {errors.invoiceClientId?.message}
                      </span>
                    )
                  }
                  error={!!errors.invoiceClientId}
                  disabled={loading}
                />
              );
            }}
          />
        )}
      />

      <Typography variant="h6" sx={{ mt: 2 }}>
        Invoice items
      </Typography>

      <Grid
        container
        spacing={3}
        sx={{ mt: 0, mb: 1 }}
        data-test-id="invoice-item-1"
      >
        <Grid item xs={12} sm={6}>
          <ControlledFormTextField
            sx={{ m: 0 }}
            id="items.0.description"
            control={control}
            label="Description"
            errorField={errors.items?.at(0)?.description}
            dataTestId="invoice-item-description"
            disabled={loading}
            required
          />
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormTextField
            sx={{ m: 0 }}
            id="items.0.value"
            label="Value"
            errorField={errors.items?.at(0)?.value}
            dataTestId="invoice-item-value"
            register={register}
            disabled={loading}
            type="number"
            required
          />
        </Grid>
      </Grid>

      {fields.slice(1).map((field, index) => {
        // need to offset index since first item is fixed
        const idx = index + 1;
        return (
          <div key={field.id}>
            <Divider sx={{ mt: 3 }} />
            <Grid
              container
              spacing={3}
              sx={{ mt: 0, mb: 1 }}
              data-test-id={`invoice-item-${idx + 1}`}
            >
              <Grid item xs={12} sm={6}>
                <ControlledFormTextField
                  sx={{ m: 0 }}
                  id={`items.${idx}.description`}
                  control={control}
                  label="Description"
                  errorField={errors.items?.at(idx)?.description}
                  dataTestId={`invoice-item-description`}
                  disabled={loading}
                  required
                />
              </Grid>
              <Grid item xs={12} sm={6}>
                <Stack direction="row" spacing={3}>
                  <FormTextField
                    sx={{ m: 0 }}
                    id={`items.${idx}.value`}
                    label="Value"
                    errorField={errors.items?.at(idx)?.value}
                    dataTestId={`invoice-item-value`}
                    register={register}
                    disabled={loading}
                    type="number"
                    required
                  />
                  <IconButton
                    onClick={() => remove(idx)}
                    sx={{ minWidth: 56 }}
                    disabled={loading}
                  >
                    <ClearIcon />
                  </IconButton>
                </Stack>
              </Grid>
            </Grid>
          </div>
        );
      })}

      <Button
        variant="outlined"
        fullWidth
        size="large"
        sx={{ mt: 2, mb: 1 }}
        onClick={() => append({ description: "", value: undefined })}
        disabled={loading}
      >
        Add
      </Button>

      <Box sx={{ display: "flex", alignItems: "baseline" }}>
        <Typography variant="h5" sx={{ flexGrow: 1 }}>
          Total
        </Typography>

        <Typography variant="h5">$ &nbsp;</Typography>

        <Typography variant="h4">{total.toFixed(2)}</Typography>
      </Box>

      <LoadingButton
        sx={{ mt: 2, mb: 2 }}
        type="submit"
        dataTestId="submit-client"
        loading={loading}
      >
        {isUpdateForm ? "Update" : "Register"}
      </LoadingButton>
    </Box>
  );
}
