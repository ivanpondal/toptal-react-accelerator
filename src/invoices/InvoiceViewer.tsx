import { Alert, Box, Divider, Typography } from "@mui/material";

export type InvoiceDetailsFormData = {
  invoiceDate: Date;
  invoiceDueDate: Date;
  invoiceNumber: string;
  invoiceProjectCode?: string;
  invoiceClientId: string;
  items: Array<{
    description: string;
    value: number;
  }>;
  total: number;
};

export type InvoiceViewerProps = {
  invoice?: InvoiceDetailsFormData;
  clientNames?: { [key: string]: { id: string; label: string } };
  loading?: boolean;
  errorMessage?: string;
  successMessage?: string;
};

export default function InvoiceForm(props: InvoiceViewerProps) {
  const { clientNames, loading, errorMessage, invoice } = props;

  let clientName;
  if (clientNames && invoice?.invoiceClientId) {
    clientName = clientNames[invoice.invoiceClientId].label;
  }

  const total = invoice?.items
    .map((item) => item.value)
    .map((value) => Number(value))
    .filter((value) => !Number.isNaN(value))
    .filter((value) => value > 0)
    .reduce((prev, curr) => prev + curr, 0);

  return (
    <Box sx={{ mt: 1 }}>
      {errorMessage && (
        <Alert data-test="form-error" severity="error">
          {errorMessage}
        </Alert>
      )}

      <Typography variant="h2" sx={{ mt: 2 }} data-test="invoice-number">
        Invoice # {invoice?.invoiceNumber}
      </Typography>

      <Typography
        variant="h5"
        sx={{ mt: 2, textAlign: "right" }}
        data-test="invoice-date"
      >
        Date: {invoice?.invoiceDate.toLocaleDateString()}
      </Typography>

      <Typography
        variant="h5"
        sx={{ mt: 2, textAlign: "right" }}
        data-test="invoice-due-date"
      >
        Due Date: {invoice?.invoiceDueDate.toLocaleDateString()}
      </Typography>

      <Typography variant="h4" sx={{ mt: 2, textAlign: "right" }}>
        Billed to {clientName}
      </Typography>

      {invoice?.invoiceProjectCode ? (
        <Typography
          variant="h6"
          sx={{ mt: 2, textAlign: "right" }}
          data-test="invoice-project-code"
        >
          "{invoice?.invoiceProjectCode}"
        </Typography>
      ) : null}

      <Divider sx={{ mt: 3 }} />

      <Typography variant="h5" sx={{ mt: 2 }}>
        {" "}
        Invoice items
      </Typography>

      <Box sx={{ display: "flex", alignItems: "baseline", mt: 4 }}>
        <Typography variant="button" sx={{ flexGrow: 1 }}>
          Description
        </Typography>

        <Typography variant="button">Value</Typography>
      </Box>

      {invoice?.items.map((item, index) => (
        <div key={index} data-test={`invoice-item-${index + 1}`}>
          <Divider sx={{ mt: 3 }} />
          <Box sx={{ display: "flex", alignItems: "baseline" }}>
            <Typography
              variant="subtitle1"
              sx={{ flexGrow: 1 }}
              data-test="invoice-item-description"
            >
              {item.description}
            </Typography>

            <Typography
              variant="h6"
              sx={{ mt: 2 }}
              data-test="invoice-item-value"
            >
              {item.value}
            </Typography>
          </Box>
        </div>
      ))}

      <Box sx={{ display: "flex", alignItems: "baseline", mt: 6 }}>
        <Typography variant="h5" sx={{ flexGrow: 1 }}>
          Total
        </Typography>

        <Typography variant="h5">$ &nbsp;</Typography>

        <Typography variant="h4" data-test="invoice-total">
          {total}
        </Typography>
      </Box>
    </Box>
  );
}
