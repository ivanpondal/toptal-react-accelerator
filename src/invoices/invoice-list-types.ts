const invoiceSortingFields = [
  "total",
  "dueDate",
  "creationDate",
  "companyName",
] as const;
type InvoiceSortingFieldTuple = typeof invoiceSortingFields;
export type InvoiceSortingField = InvoiceSortingFieldTuple[number];

export function isInvoiceSortingField(
  value: string
): value is InvoiceSortingField {
  return invoiceSortingFields.includes(value as InvoiceSortingField);
}

const sortingOrders = ["asc", "desc"] as const;
type sortingOrderTuple = typeof sortingOrders;
export type SortingOrder = sortingOrderTuple[number];

export function isSortingOrder(value: string): value is SortingOrder {
  return sortingOrders.includes(value as SortingOrder);
}

export type InvoiceSortingParams = {
  field?: InvoiceSortingField;
  order?: SortingOrder;
};
