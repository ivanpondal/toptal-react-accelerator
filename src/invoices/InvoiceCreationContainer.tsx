import { ClientAPI, InvoiceAPI } from "../api/base";
import InvoiceForm, { InvoiceDetailsFormData } from "./InvoiceForm";
import { useAsync } from "../hooks/useAsync";
import { useCallback, useEffect, useState } from "react";

export default function InvoiceCreationContainer() {
  const { execute: loadClientNames, value: clientNames } = useAsync(
    useCallback(async () => {
      return await ClientAPI.getAllNames()
        .then((res) => res.clients)
        .then((clients) =>
          clients.map((item) => [
            item.id,
            {
              id: item.id,
              label: item.companyName,
            },
          ])
        )
        .then((clientPairs) => Object.fromEntries(clientPairs));
    }, [])
  );

  const [successfulUpdateMessage, setSuccessfulUpdateMessage] = useState<
    string | null
  >(null);

  const { execute, value, error, status } = useAsync(
    useCallback(async (params: InvoiceDetailsFormData) => {
      setSuccessfulUpdateMessage(null);
      return await InvoiceAPI.create({
        invoice_number: params.invoiceNumber,
        client_id: params.invoiceClientId,
        date: params.invoiceDate.valueOf(),
        dueDate: params.invoiceDueDate.valueOf(),
        value: params.total,
        projectCode: params.invoiceProjectCode,
        meta: { items: params.items },
      }).then(() =>
        setSuccessfulUpdateMessage("Invoice created successfully!")
      );
    }, [])
  );

  useEffect(() => {
    loadClientNames({});
  }, [loadClientNames]);

  let clientNamesForm = clientNames ? clientNames : {};

  let errorMessage;
  if (error === "invoice with that number already exists") {
    errorMessage = "Invoice number already exists";
  } else if (error) {
    errorMessage = "Oops! Something went wrong with the server";
  }

  let successMessage = successfulUpdateMessage
    ? successfulUpdateMessage
    : undefined;

  return (
    <InvoiceForm
      onSubmitRequest={execute}
      clientNames={clientNamesForm}
      loading={status === "pending"}
      errorMessage={errorMessage}
      successMessage={successMessage}
    />
  );
}
