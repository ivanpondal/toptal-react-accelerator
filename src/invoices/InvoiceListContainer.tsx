import { Alert, Box, IconButton, Typography } from "@mui/material";
import { InvoicesTable } from "./InvoicesTable";
import { useRouter } from "next/router";
import { useEffect } from "react";
import AddBoxIcon from "@mui/icons-material/AddBoxOutlined";
import Link from "next/link";
import { useInvoiceStore } from "./InvoiceStore";
import { InvoiceSortingParams } from "./invoice-list-types";

export const InvoiceListContainer = (props: {
  sorting?: InvoiceSortingParams;
}) => {
  const { sorting } = props;

  const tableSortModel = sorting
    ? [
        {
          field: sorting.field as string,
          sort: sorting.order,
        },
      ]
    : [];

  const router = useRouter();

  const { invoices, fetchStatus, errorMessage } = useInvoiceStore(
    (state) => state.invoiceList
  );
  const fetchInvoiceList = useInvoiceStore((state) => state.fetchInvoiceList);

  useEffect(() => {
    fetchInvoiceList({ sort: sorting });
  }, [fetchInvoiceList, sorting]);

  return (
    <div style={{ display: "block" }}>
      {errorMessage && (
        <Alert sx={{ mb: 1 }} severity="error" data-test="invoices-fetch-error">
          {errorMessage}
        </Alert>
      )}

      <Box sx={{ display: "flex", mt: 3, mb: 1 }}>
        <Typography variant="h5" sx={{ flexGrow: 1 }}>
          Invoices
        </Typography>
        <Link href="/invoices/new" passHref>
          <IconButton data-test="add-invoice">
            <AddBoxIcon color="primary" />
          </IconButton>
        </Link>
      </Box>
      <InvoicesTable
        invoices={invoices}
        loading={fetchStatus === "pending"}
        onRowClick={(rowId) => router.push(`/invoices/${rowId}/view`)}
        sortable={true}
        sortModel={tableSortModel}
        onSortModelChange={(model) => {
          if (model.length === 0) {
            router.push("/invoices");
          } else {
            router.push(
              `/invoices?sortBy=${
                model[0].field
              }&sortOrder=${model[0].sort?.toUpperCase()}`
            );
          }
        }}
      />
    </div>
  );
};
