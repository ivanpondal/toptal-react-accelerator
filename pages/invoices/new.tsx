import { Container, Box, Typography } from "@mui/material";
import { AuthGuard } from "../../src/auth/AuthGuard";
import ClientCreationContainer from "../../src/clients/ClientCreationContainer";
import { CompanyDetailsGuard } from "../../src/company/CompanyDetailsGuard";
import InvoiceCreationContainer from "../../src/invoices/InvoiceCreationContainer";

export default function NewInvoice() {
  return (
    <AuthGuard>
      <CompanyDetailsGuard>
        <Container component="main" maxWidth="sm">
          <Box
            sx={{
              marginTop: 8,
              display: "flex",
              flexDirection: "column",
              alignItems: "center",
            }}
          >
            <Typography variant="h5">
              Create invoice
            </Typography>
            <InvoiceCreationContainer />
          </Box>
        </Container>
      </CompanyDetailsGuard>
    </AuthGuard>
  );
}
