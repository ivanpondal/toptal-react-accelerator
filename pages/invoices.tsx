import { Container, Box, Typography, Grid } from "@mui/material";
import { AuthGuard } from "../src/auth/AuthGuard";
import { CompanyDetailsGuard } from "../src/company/CompanyDetailsGuard";
import { InvoiceListContainer } from "../src/invoices/InvoiceListContainer";

export default function InvoiceList() {
  return (
    <AuthGuard>
      <CompanyDetailsGuard>
        <Container component="main" maxWidth="lg">
          <InvoiceListContainer />
        </Container>
      </CompanyDetailsGuard>
    </AuthGuard>
  );
}
