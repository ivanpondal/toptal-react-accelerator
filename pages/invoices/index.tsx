import { Container } from "@mui/material";
import { useRouter } from "next/router";
import { useEffect, useState } from "react";
import { AuthGuard } from "../../src/auth/AuthGuard";
import { CompanyDetailsGuard } from "../../src/company/CompanyDetailsGuard";
import NavBarContainer from "../../src/components/NavBarContainer";
import parseQueryParam from "../../src/integration/query-params";
import {
  InvoiceSortingParams,
  InvoiceSortingField,
  isInvoiceSortingField,
  SortingOrder,
  isSortingOrder,
} from "../../src/invoices/invoice-list-types";
import { InvoiceListContainer } from "../../src/invoices/InvoiceListContainer";

export default function InvoiceList() {
  const router = useRouter();

  const [sortingParams, setSortingParams] = useState<
    InvoiceSortingParams | undefined
  >(undefined);
  // router sorting params change event
  useEffect(() => {
    if (router.isReady) {
      if (router.query.sortBy) {
        const parsedSortBy = parseQueryParam(router.query.sortBy);

        let sortingField: InvoiceSortingField | undefined;
        if (parsedSortBy && isInvoiceSortingField(parsedSortBy)) {
          sortingField = parsedSortBy;
        }

        const parsedSortOrder = parseQueryParam(
          router.query.sortOrder
        )?.toLowerCase();

        let sortingOrder: SortingOrder | undefined;
        if (parsedSortOrder && isSortingOrder(parsedSortOrder)) {
          sortingOrder = parsedSortOrder;
        }

        if (sortingField) {
          setSortingParams({
            field: sortingField,
            order: sortingOrder,
          });
        }
      } else {
        setSortingParams(undefined);
      }
    }
  }, [router.isReady, router.query.sortBy, router.query.sortOrder]);

  const [page, setPage] = useState<number | undefined>(undefined);
  // router paging params change event
  useEffect(() => {
    if (router.isReady) {
      if (router.query.page) {
        const parsedPage = parseQueryParam(router.query.page);
        const pageNumber = parsedPage ? parseInt(parsedPage) : NaN;
        if (!isNaN(pageNumber) && pageNumber > 0) {
          setPage(pageNumber);
        }
      } else {
        setPage(undefined);
      }
    }
  }, [router.isReady, router.query.page]);

  const [companyNameFilter, setCompanyNameFilter] = useState<
    string | undefined
  >(undefined);
  // router filtering params change event
  useEffect(() => {
    if (router.isReady) {
      if (router.query.companyFilter) {
        const companyName = parseQueryParam(router.query.companyFilter);
        if (companyName) {
          setCompanyNameFilter(companyName);
        }
      } else {
        setCompanyNameFilter(undefined);
      }
    }
  }, [router.isReady, router.query.companyFilter]);

  return (
    <AuthGuard>
      <NavBarContainer activePage="invoices" />
      <CompanyDetailsGuard>
        <Container component="main" maxWidth="lg">
          <InvoiceListContainer
            sorting={sortingParams}
            page={page}
            companyNameFilter={companyNameFilter}
          />
        </Container>
      </CompanyDetailsGuard>
    </AuthGuard>
  );
}
