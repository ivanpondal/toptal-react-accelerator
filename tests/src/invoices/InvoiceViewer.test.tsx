import renderer from "react-test-renderer";
import { InvoiceDetailsFormData } from "../../../src/invoices/InvoiceForm";
import InvoiceViewer from "../../../src/invoices/InvoiceViewer";

const invoiceWithoutProjectCode: InvoiceDetailsFormData = {
  invoiceDate: new Date("01 Jan 2022 00:00:00 GMT"),
  invoiceDueDate: new Date("10 Jan 2022 00:00:00 GMT"),
  invoiceNumber: "001",
  invoiceClientId: "1",
  items: [{ description: "Fertilizer", value: 10000 }],
  total: 10000,
};

const clientNames = {
  "1": {
    id: "1",
    label: "Fertilot",
  },
};

describe("invoice and client names props", () => {
  it("should render as invoice and billing data", () => {
    const tree = renderer
      .create(
        <InvoiceViewer
          invoice={invoiceWithoutProjectCode}
          clientNames={clientNames}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("invoice with project code and client names props", () => {
  it("should render as invoice and billing data with project code", () => {
    const invoiceWithProjectCode = {
      ...invoiceWithoutProjectCode,
      invoiceProjectCode: "cool project",
    };

    const tree = renderer
      .create(
        <InvoiceViewer
          invoice={invoiceWithProjectCode}
          clientNames={clientNames}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
