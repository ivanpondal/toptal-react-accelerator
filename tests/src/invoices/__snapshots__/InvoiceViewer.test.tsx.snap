// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invoice and client names props should render as invoice and billing data 1`] = `
<div
  className="MuiBox-root css-ru8kxu"
>
  <h2
    className="MuiTypography-root MuiTypography-h2 invoiceHeader css-1ckgmu4-MuiTypography-root"
    data-test="invoice-number"
  >
    Invoice # 
    001
  </h2>
  <h5
    className="MuiTypography-root MuiTypography-h5 invoiceHeader css-xh7455-MuiTypography-root"
    data-test="invoice-date"
  >
    Date: 
    12/31/2021
  </h5>
  <h5
    className="MuiTypography-root MuiTypography-h5 invoiceHeader css-xh7455-MuiTypography-root"
    data-test="invoice-due-date"
  >
    Due Date: 
    1/9/2022
  </h5>
  <h4
    className="MuiTypography-root MuiTypography-h4 invoiceHeader css-1t413ox-MuiTypography-root"
  >
    Billed to 
    Fertilot
  </h4>
  <hr
    className="MuiDivider-root MuiDivider-fullWidth css-1sjh9xo-MuiDivider-root"
  />
  <h5
    className="MuiTypography-root MuiTypography-h5 css-1657z6w-MuiTypography-root"
  >
     
    Invoice items
  </h5>
  <div
    className="MuiBox-root css-1fnf33c"
  >
    <span
      className="MuiTypography-root MuiTypography-button css-1o0kh3y-MuiTypography-root"
    >
      Description
    </span>
    <span
      className="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
    >
      Value
    </span>
  </div>
  <div
    className="jsx-68126bc088c3b37f"
    data-test="invoice-item-1"
  >
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-1sjh9xo-MuiDivider-root"
    />
    <div
      className="MuiBox-root css-x1sij0"
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle1 css-1c1mbpd-MuiTypography-root"
        data-test="invoice-item-description"
      >
        Fertilizer
      </h6>
      <h6
        className="MuiTypography-root MuiTypography-h6 css-1ry679h-MuiTypography-root"
        data-test="invoice-item-value"
      >
        10000
      </h6>
    </div>
  </div>
  <div
    className="MuiBox-root css-ynyiwi"
  >
    <h5
      className="MuiTypography-root MuiTypography-h5 css-1f8a19r-MuiTypography-root"
    >
      Total
    </h5>
    <h5
      className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
    >
      $  
    </h5>
    <h4
      className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      data-test="invoice-total"
    >
      10000
    </h4>
  </div>
</div>
`;

exports[`invoice with project code and client names props should render as invoice and billing data with project code 1`] = `
<div
  className="MuiBox-root css-ru8kxu"
>
  <h2
    className="MuiTypography-root MuiTypography-h2 invoiceHeader css-1ckgmu4-MuiTypography-root"
    data-test="invoice-number"
  >
    Invoice # 
    001
  </h2>
  <h5
    className="MuiTypography-root MuiTypography-h5 invoiceHeader css-xh7455-MuiTypography-root"
    data-test="invoice-date"
  >
    Date: 
    12/31/2021
  </h5>
  <h5
    className="MuiTypography-root MuiTypography-h5 invoiceHeader css-xh7455-MuiTypography-root"
    data-test="invoice-due-date"
  >
    Due Date: 
    1/9/2022
  </h5>
  <h4
    className="MuiTypography-root MuiTypography-h4 invoiceHeader css-1t413ox-MuiTypography-root"
  >
    Billed to 
    Fertilot
  </h4>
  <h6
    className="MuiTypography-root MuiTypography-h6 invoiceHeader css-1nrmkr4-MuiTypography-root"
    data-test="invoice-project-code"
  >
    "
    cool project
    "
  </h6>
  <hr
    className="MuiDivider-root MuiDivider-fullWidth css-1sjh9xo-MuiDivider-root"
  />
  <h5
    className="MuiTypography-root MuiTypography-h5 css-1657z6w-MuiTypography-root"
  >
     
    Invoice items
  </h5>
  <div
    className="MuiBox-root css-1fnf33c"
  >
    <span
      className="MuiTypography-root MuiTypography-button css-1o0kh3y-MuiTypography-root"
    >
      Description
    </span>
    <span
      className="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
    >
      Value
    </span>
  </div>
  <div
    className="jsx-68126bc088c3b37f"
    data-test="invoice-item-1"
  >
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-1sjh9xo-MuiDivider-root"
    />
    <div
      className="MuiBox-root css-x1sij0"
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle1 css-1c1mbpd-MuiTypography-root"
        data-test="invoice-item-description"
      >
        Fertilizer
      </h6>
      <h6
        className="MuiTypography-root MuiTypography-h6 css-1ry679h-MuiTypography-root"
        data-test="invoice-item-value"
      >
        10000
      </h6>
    </div>
  </div>
  <div
    className="MuiBox-root css-ynyiwi"
  >
    <h5
      className="MuiTypography-root MuiTypography-h5 css-1f8a19r-MuiTypography-root"
    >
      Total
    </h5>
    <h5
      className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
    >
      $  
    </h5>
    <h4
      className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      data-test="invoice-total"
    >
      10000
    </h4>
  </div>
</div>
`;
